catalog/security/catalog-security-operationplugin/src/main/java/ddf/catalog/security/operation/plugin/OperationPlugin.java

```diff
@@ -0,0 +1,113 @@
+import ddf.catalog.operation.QueryResponse;
+import ddf.catalog.operation.UpdateRequest;
+import ddf.catalog.plugin.AccessPlugin;
+import ddf.catalog.plugin.PolicyPlugin;
+import ddf.catalog.plugin.StopProcessingException;
+import ddf.security.SecurityConstants;
+import ddf.security.Subject;
+import ddf.security.SubjectUtils;
+import ddf.security.permission.CollectionPermission;
+import ddf.security.permission.KeyValueCollectionPermission;
+
+/**
+ * Security-based plugin that check the pre create/update/delete/query request to make sure the
+ * subject has the appropriate attributes to make the request.
+ */
+public class OperationPlugin implements AccessPlugin {
+
+    @Override
+    public CreateRequest processPreCreate(CreateRequest input) throws StopProcessingException {
+        checkOperation(input);
+        return input;
+    }
+
+    @Override
+    public UpdateRequest processPreUpdate(UpdateRequest input) throws StopProcessingException {
+        checkOperation(input);
+        return input;
+    }
+
+    @Override
+    public DeleteRequest processPreDelete(DeleteRequest input) throws StopProcessingException {
+        checkOperation(input);
+        return input;
+    }
+
+    @Override
+    public QueryRequest processPreQuery(QueryRequest input) throws StopProcessingException {
+        checkOperation(input);
+        return input;
+    }
+
+    @Override
+    public QueryResponse processPostQuery(QueryResponse input) throws StopProcessingException {
+        return input;
+    }
+
+    /**
+     * checkOperation will throw a StopProcessingException if the operation is not permitted
+     * based on the the subjects attributes and the operations property "operation.security"
+     *
+     * @param operation
+     * @throws StopProcessingException
+     */
+    private void checkOperation(Operation operation) throws StopProcessingException {
+        if (!operation.hasProperties()
+                || !operation.containsPropertyName(PolicyPlugin.OPERATION_SECURITY)) {
+            return;
+        }
+
+        Object securityAssertion = operation.getPropertyValue(SecurityConstants.SECURITY_SUBJECT);
+        Subject subject;
+        if (securityAssertion instanceof Subject) {
+            subject = (Subject) securityAssertion;
+        } else {
+            throw new StopProcessingException(
+                    "Unable to filter contents of current message, no user Subject available.");
+        }
+
+        KeyValueCollectionPermission securityPermission = new KeyValueCollectionPermission(
+                CollectionPermission.READ_ACTION);
+
+        Map<String, Set<String>> perms = (Map<String, Set<String>>) operation.getPropertyValue(
+                PolicyPlugin.OPERATION_SECURITY);
+        Map<String, List<String>> kvPerms = new HashMap<String, List<String>>();
+        for (Map.Entry<String, Set<String>> entry : perms.entrySet()) {
+            kvPerms.put(entry.getKey(), new ArrayList<String>(entry.getValue()));
+        }
```

Comment:
It looks like we just need to add a new constructor for KeyValueCollectionPermission to me. I don't want someone to need to do this for every single access plugin. There is no reason why we can't just update that impl and remove all this code.